File: historical_pull.R
#!/usr/bin/env Rscript
# historical_pull.R -- R port of historical_pull.py, for running from a
# network unrestricted from api.the-odds-api.com (Claude's remote sandbox
# for this project is not, per its org's egress policy).
#
# Does the pull AND the flatten in one pass: for each week-1 FBS-vs-FBS
# game, pulls the closing-line historical player-prop odds, checkpoints
# the raw JSON to hist_raw/ (so a crash/rerun never re-spends), and writes
# one tidy CSV -- one row per (game, player, market) with the consensus
# closing line and median over/under price across books. Upload that CSV
# back; no need to upload hist_raw/ itself.
#
# SETUP
#   install.packages(c("httr", "jsonlite"))          # if not already installed
#   Sys.setenv(ODDS_API_KEY = "your_real_key")        # do NOT hardcode it below
#   (or put ODDS_API_KEY=... in a .Renviron in this folder and restart R)
#
# Run from the folder containing 2025_schedule.csv and team_map.csv:
#   Rscript historical_pull.R --week 1 --dry-run     # spends nothing, sanity check
#   Rscript historical_pull.R --week 1               # the real pull
#
# --game-lines pulls spreads+totals instead of player props (a separate,
# much cheaper pull -- 2 markets instead of 6) -- for the game-context
# volume adjustment (implied team total / spread -> expected pace and
# rush/pass split). Writes hist_lines_closing_wkN.csv, one row per game:
# game_id, week, home_team, away_team, home_spread, total. Checkpoints to
# a separate hist_raw_lines/ dir so it can't collide with a player-prop
# pull's cache.
#   Rscript historical_pull.R --week-start 1 --week-end 15 --game-lines --dry-run
#   Rscript historical_pull.R --week-start 1 --week-end 15 --game-lines
#
# COST MODEL (matches historical_pull.py, same non-negotiable math):
#   historical event-odds = 10 credits x regions x markets x event x snapshot.
#   Closing-only (1 snapshot), 1 region => 60 credits/game for the 6-market
#   player-prop pull, 20 credits/game for the 2-market --game-lines pull.

suppressMessages({
  library(httr)
  library(jsonlite)
})

# Team/player names in the source files have accents (San Jose State) and
# non-ASCII chars (Hawai'i); force a UTF-8 locale so string handling below
# doesn't silently mangle them on a system that defaults to something else.
invisible(try(Sys.setlocale("LC_CTYPE", "C.UTF-8"), silent = TRUE))

# ----------------- CONFIG -----------------
API_KEY <- Sys.getenv("ODDS_API_KEY")
if (!nzchar(API_KEY)) {
  stop("ODDS_API_KEY is not set. Run Sys.setenv(ODDS_API_KEY=\"...\") first (or set it in .Renviron) -- do not hardcode it in this file.")
}
SPORT    <- "americanfootball_ncaaf"
REGION   <- "us"
MARKETS  <- c("player_pass_yds", "player_pass_attempts",
              "player_rush_yds", "player_rush_attempts",
              "player_reception_yds", "player_receptions")
ODDS_FMT <- "american"

HIST_COST_PER  <- 10
CREDIT_CEILING <- 100000
CREDIT_FLOOR   <- 2000
BASE <- "https://api.the-odds-api.com/v4"

SCRIPT_DIR  <- dirname(sub("--file=", "", grep("--file=", commandArgs(trailingOnly = FALSE), value = TRUE)))
if (length(SCRIPT_DIR) == 0 || !nzchar(SCRIPT_DIR)) SCRIPT_DIR <- "."
# CKPT_DIR is set after arg parsing below (depends on --game-lines).
SNAPSHOTS   <- c("closing")            # closing-only, per PROJECT_STATE.md decision
SNAPSHOT_OFFSET_HOURS <- c(opening = 36, closing = 1)

FBS_CONF <- c("SEC", "Big Ten", "Big 12", "ACC", "American Athletic",
              "Mountain West", "Sun Belt", "Conference USA", "Mid-American",
              "Pac-12", "FBS Independents")

# ----------------- ARGS -----------------
parse_args <- function(argv) {
  a <- list(schedule = file.path(SCRIPT_DIR, "2025_schedule.csv"),
            team_map = file.path(SCRIPT_DIR, "team_map.csv"),
            season = 2025, week = NA, week_start = NA, week_end = NA,
            season_type = "regular",
            max_games = NA, include_one_fbs = FALSE, dry_run = FALSE,
            game_lines = FALSE, out = NA)
  i <- 1
  while (i <= length(argv)) {
    k <- argv[i]
    val <- function() { i <<- i + 1; argv[i] }
    if (k == "--schedule") a$schedule <- val()
    else if (k == "--team-map") a$team_map <- val()
    else if (k == "--season") a$season <- as.integer(val())
    else if (k == "--week") a$week <- as.integer(val())
    else if (k == "--week-start") a$week_start <- as.integer(val())
    else if (k == "--week-end") a$week_end <- as.integer(val())
    else if (k == "--season-type") a$season_type <- val()
    else if (k == "--max-games") a$max_games <- as.integer(val())
    else if (k == "--include-one-fbs") a$include_one_fbs <- TRUE
    else if (k == "--dry-run") a$dry_run <- TRUE
    else if (k == "--game-lines") a$game_lines <- TRUE
    else if (k == "--out") a$out <- val()
    else stop(paste("unknown arg:", k))
    i <- i + 1
  }
  if (xor(is.na(a$week_start), is.na(a$week_end)))
    stop("--week-start and --week-end must be given together")
  if (!is.na(a$week) && !is.na(a$week_start))
    stop("use --week for a single week OR --week-start/--week-end for a range, not both")
  if (is.na(a$out)) {
    wk <- if (!is.na(a$week_start)) paste0(a$week_start, "-", a$week_end)
          else if (is.na(a$week)) "all" else a$week
    prefix <- if (a$game_lines) "hist_lines_closing_wk" else "hist_props_closing_wk"
    a$out <- file.path(SCRIPT_DIR, paste0(prefix, wk, ".csv"))
  }
  a
}
args <- parse_args(commandArgs(trailingOnly = TRUE))

# --game-lines pulls spreads+totals (2 markets) instead of the 6 player-prop
# markets, and checkpoints separately so the two pulls' caches can't collide.
if (args$game_lines) {
  MARKETS <- c("spreads", "totals")
  CKPT_DIR <- file.path(SCRIPT_DIR, "hist_raw_lines")
} else {
  CKPT_DIR <- file.path(SCRIPT_DIR, "hist_raw")
}

# ----------------- SCHEDULE -----------------
load_schedule <- function(path, season, season_type, fbs_only = TRUE, both_fbs = TRUE) {
  df <- read.csv(path, stringsAsFactors = FALSE, fileEncoding = "UTF-8",
                 colClasses = "character")
  names(df) <- trimws(names(df))
  names(df)[1] <- sub("^﻿", "", names(df)[1])         # strip a leading BOM if the reader left one
  df <- df[!duplicated(df$Id), ]                          # dedup: ~87 rows repeated per broadcast outlet
  df <- df[df$Season == as.character(season), ]
  if (season_type != "all") df <- df[df$SeasonType == season_type, ]
  if (fbs_only) {
    h_fbs <- df$HomeConference %in% FBS_CONF
    a_fbs <- df$AwayConference %in% FBS_CONF
    df <- if (both_fbs) df[h_fbs & a_fbs, ] else df[h_fbs | a_fbs, ]
  }
  df <- df[nzchar(df$StartTime), ]
  df$commence_dt <- as.POSIXct(df$StartTime, format = "%Y-%m-%dT%H:%M:%OSZ", tz = "UTC")
  df <- df[order(df$commence_dt), ]
  df
}

snapshot_iso <- function(commence_dt, label) {
  ts <- commence_dt - SNAPSHOT_OFFSET_HOURS[[label]] * 3600
  format(ts, "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
}

# ----------------- TEAM NAME MATCHING -----------------
norm_name <- function(s) {
  s <- iconv(s, from = "UTF-8", to = "ASCII//TRANSLIT")     # strip accents (e.g. Jose -> Jose)
  s <- tolower(ifelse(is.na(s), "", s))
  gsub("[^a-z0-9]", "", s)
}

load_team_map <- function(path) {
  if (!file.exists(path)) return(list())
  df <- read.csv(path, stringsAsFactors = FALSE, fileEncoding = "UTF-8")
  setNames(as.list(norm_name(df$odds_name)), norm_name(df$cfbd_name))
}
team_map <- load_team_map(args$team_map)

apply_map <- function(team) {
  n <- norm_name(team)
  if (!is.null(team_map[[n]])) team_map[[n]] else n
}

match_event <- function(events, home, away) {
  h <- apply_map(home); a <- apply_map(away)
  for (e in events) {
    eh <- norm_name(e$home_team %||% ""); ea <- norm_name(e$away_team %||% "")
    fwd <- (grepl(h, eh, fixed = TRUE) || grepl(eh, h, fixed = TRUE)) &&
           (grepl(a, ea, fixed = TRUE) || grepl(ea, a, fixed = TRUE))
    rev <- (grepl(h, ea, fixed = TRUE) || grepl(ea, h, fixed = TRUE)) &&
           (grepl(a, eh, fixed = TRUE) || grepl(eh, a, fixed = TRUE))
    if (fwd || rev) return(e)
  }
  NULL
}
`%||%` <- function(a, b) if (is.null(a)) b else a

# ----------------- HTTP -----------------
api_get <- function(path, query) {
  resp <- tryCatch(
    GET(paste0(BASE, path), query = c(query, apiKey = API_KEY), timeout(40)),
    error = function(e) { message("  error: ", conditionMessage(e)); NULL }
  )
  if (is.null(resp)) return(list(body = NULL, remaining = NA))
  remaining <- headers(resp)[["x-requests-remaining"]]
  if (http_error(resp)) {
    message("  HTTP ", status_code(resp), ": ", substr(content(resp, "text", encoding = "UTF-8"), 1, 200))
    return(list(body = NULL, remaining = if (is.null(remaining)) NA else as.integer(remaining)))
  }
  body <- fromJSON(content(resp, "text", encoding = "UTF-8"), simplifyVector = FALSE)
  list(body = body, remaining = if (is.null(remaining)) NA else as.integer(remaining))
}

hist_events <- function(date_iso) {
  r <- api_get(paste0("/historical/sports/", SPORT, "/events"), list(date = date_iso))
  events <- if (!is.null(r$body$data)) r$body$data else r$body
  list(events = events, remaining = r$remaining)
}

hist_event_odds <- function(event_id, date_iso) {
  api_get(paste0("/historical/sports/", SPORT, "/events/", event_id, "/odds"),
          list(date = date_iso, regions = REGION, markets = paste(MARKETS, collapse = ","),
               oddsFormat = ODDS_FMT))
}

# ----------------- CHECKPOINT -----------------
ckpt_path <- function(event_id, label) file.path(CKPT_DIR, paste0(event_id, "_", label, ".json"))
already_have <- function(event_id, label) {
  p <- ckpt_path(event_id, label)
  file.exists(p) && file.info(p)$size > 2
}
save_ckpt <- function(event_id, label, body) {
  dir.create(CKPT_DIR, showWarnings = FALSE, recursive = TRUE)
  writeLines(toJSON(body, auto_unbox = TRUE, null = "null"), ckpt_path(event_id, label))
}
read_ckpt <- function(event_id, label) {
  fromJSON(ckpt_path(event_id, label), simplifyVector = FALSE)
}

# ----------------- FLATTEN -----------------
# One snapshot body -> long rows (game, book, market, player, side, line, price),
# then collapsed to one row per (game, player, market) with the median line and
# median over/under price across books.
flatten_snapshot <- function(body, week) {
  ev <- body$data
  if (is.null(ev) || is.null(ev$bookmakers)) return(NULL)
  rows <- list()
  for (bk in ev$bookmakers) {
    for (mkt in bk$markets %||% list()) {
      for (oc in mkt$outcomes %||% list()) {
        rows[[length(rows) + 1]] <- data.frame(
          game_id = ev$id, week = week,
          home_team = ev$home_team %||% NA, away_team = ev$away_team %||% NA,
          commence_time = ev$commence_time %||% NA,
          book = bk$key %||% NA, market = mkt$key %||% NA,
          player = oc$description %||% NA, side = oc$name %||% NA,
          line = as.numeric(oc$point %||% NA), price = as.numeric(oc$price %||% NA),
          stringsAsFactors = FALSE
        )
      }
    }
  }
  if (length(rows) == 0) return(NULL)
  do.call(rbind, rows)
}

consensus <- function(long_df) {
  if (is.null(long_df) || nrow(long_df) == 0) return(long_df)
  key <- interaction(long_df$game_id, long_df$player, long_df$market, drop = TRUE)
  out <- do.call(rbind, lapply(split(long_df, key), function(g) {
    data.frame(
      game_id = g$game_id[1], week = g$week[1],
      home_team = g$home_team[1], away_team = g$away_team[1],
      commence_time = g$commence_time[1],
      market = g$market[1], player = g$player[1],
      book_line = median(g$line, na.rm = TRUE),
      over_price = suppressWarnings(median(g$price[g$side == "Over"], na.rm = TRUE)),
      under_price = suppressWarnings(median(g$price[g$side == "Under"], na.rm = TRUE)),
      n_books = length(unique(g$book)),
      stringsAsFactors = FALSE
    )
  }))
  rownames(out) <- NULL
  out
}

# For spreads/totals, oc$name is a TEAM NAME (spreads) or "Over"/"Under"
# (totals) -- flatten_snapshot's generic "side" column already captures
# this correctly, it's just consensus-collapsed differently: by
# (game_id, market, side) instead of (game_id, player, market), since
# there's no player here.
consensus_lines <- function(long_df) {
  if (is.null(long_df) || nrow(long_df) == 0) return(long_df)
  key <- interaction(long_df$game_id, long_df$market, long_df$side, drop = TRUE)
  out <- do.call(rbind, lapply(split(long_df, key), function(g) {
    data.frame(
      game_id = g$game_id[1], week = g$week[1],
      home_team = g$home_team[1], away_team = g$away_team[1],
      market = g$market[1], side = g$side[1],
      point = median(g$line, na.rm = TRUE),
      n_books = length(unique(g$book)),
      stringsAsFactors = FALSE
    )
  }))
  rownames(out) <- NULL
  out
}

# Pivot the long (game, market, side) consensus into one row per game:
# game_id, week, home_team, away_team, home_spread, total. home_spread is
# whichever spreads row's side matches home_team (normalized) -- the
# Odds API quotes a spreads outcome's point directly as that named team's
# own line, so no sign-flipping needed here.
widen_lines <- function(cons) {
  if (is.null(cons) || nrow(cons) == 0) return(cons)
  by_game <- split(cons, cons$game_id)
  rows <- lapply(by_game, function(g) {
    home_row  <- g[g$market == "spreads" & norm_name(g$side) == norm_name(g$home_team[1]), ]
    total_row <- g[g$market == "totals" & g$side == "Over", ]
    data.frame(
      game_id = g$game_id[1], week = g$week[1],
      home_team = g$home_team[1], away_team = g$away_team[1],
      home_spread = if (nrow(home_row)) home_row$point[1] else NA,
      total = if (nrow(total_row)) total_row$point[1] else NA,
      stringsAsFactors = FALSE
    )
  })
  out <- do.call(rbind, rows)
  rownames(out) <- NULL
  out
}

# ----------------- MAIN -----------------
games <- load_schedule(args$schedule, args$season, args$season_type,
                       fbs_only = TRUE, both_fbs = !args$include_one_fbs)
if (!is.na(args$week)) {
  games <- games[games$Week == as.character(args$week), ]
} else if (!is.na(args$week_start)) {
  wk_num <- suppressWarnings(as.integer(games$Week))
  games <- games[!is.na(wk_num) & wk_num >= args$week_start & wk_num <= args$week_end, ]
}
if (!is.na(args$max_games)) games <- head(games, args$max_games)

n_games <- nrow(games)
n_snaps <- length(SNAPSHOTS)
projected <- n_games * n_snaps * length(MARKETS) * HIST_COST_PER

cat(sprintf("Season %s: %d games in scope\n", args$season, n_games))
cat(sprintf("Markets: %d | snapshots/game: %d | region: %s\n", length(MARKETS), n_snaps, REGION))
cat(sprintf("Projected MAX credits (if none cached): %d x %d x %d x %d = %s\n",
            n_games, n_snaps, length(MARKETS), HIST_COST_PER, format(projected, big.mark = ",")))

cached <- if (dir.exists(CKPT_DIR)) {
  sum(sapply(SNAPSHOTS, function(lbl) length(list.files(CKPT_DIR, pattern = paste0("_", lbl, "\\.json$")))))
} else 0
cat(sprintf("Checkpoint dir: %s (%d snapshot file(s) already on disk)\n", CKPT_DIR, cached))

if (projected > CREDIT_CEILING) {
  cat(sprintf("\n*** PROJECTED %s EXCEEDS CEILING %s ***\n", format(projected, big.mark = ","), format(CREDIT_CEILING, big.mark = ",")))
  cat("Reduce scope (--max-games or --week) or raise CREDIT_CEILING deliberately.\n")
  if (!args$dry_run) quit(status = 1)
}

if (args$dry_run) {
  cat("\n[dry-run] No credits spent. Re-run without --dry-run to execute.\n")
  for (i in seq_len(min(3, n_games))) {
    cat(sprintf("  wk%s %s @ %s %s\n", games$Week[i], games$AwayTeam[i], games$HomeTeam[i],
               format(games$commence_dt[i], "%Y-%m-%dT%H:%M:%S%z")))
    for (lbl in SNAPSHOTS) cat(sprintf("      %-8s -> %s\n", lbl, snapshot_iso(games$commence_dt[i], lbl)))
  }
  quit(status = 0)
}

# ---- EXECUTE ----
dir.create(CKPT_DIR, showWarnings = FALSE, recursive = TRUE)
spent_calls <- 0
flat_rows <- list()
halted <- FALSE

for (i in seq_len(n_games)) {
  wk <- games$Week[i]; home <- games$HomeTeam[i]; away <- games$AwayTeam[i]
  disc_iso <- snapshot_iso(games$commence_dt[i], "closing")
  ev_res <- hist_events(disc_iso)
  ev <- match_event(ev_res$events, home, away)
  if (is.null(ev)) {
    cat(sprintf("[%d/%d] wk%s %s@%s: no event match at %s\n", i, n_games, wk, away, home, disc_iso))
    next
  }
  eid <- ev$id

  for (lbl in SNAPSHOTS) {
    if (already_have(eid, lbl)) {
      body <- read_ckpt(eid, lbl)
    } else {
      r <- hist_event_odds(eid, snapshot_iso(games$commence_dt[i], "closing"))
      if (is.null(r$body)) next
      save_ckpt(eid, lbl, r$body)
      spent_calls <- spent_calls + 1
      body <- r$body
      if (!is.na(r$remaining) && r$remaining < CREDIT_FLOOR) {
        cat(sprintf("\n*** remaining credits %d < floor %d. Halting with progress saved. ***\n", r$remaining, CREDIT_FLOOR))
        halted <- TRUE
      }
    }
    rows <- flatten_snapshot(body, wk)
    if (!is.null(rows)) flat_rows[[length(flat_rows) + 1]] <- rows
  }
  if (halted) break
  Sys.sleep(0.25)
  if (i %% 10 == 0) cat(sprintf("[%d/%d] wk%s done\n", i, n_games, wk))
}

cat(sprintf("\nComplete. Pulled %d new (event,snapshot) odds file(s) into %s\n", spent_calls, CKPT_DIR))

if (length(flat_rows) > 0) {
  long_df <- do.call(rbind, flat_rows)
  if (args$game_lines) {
    cons <- widen_lines(consensus_lines(long_df))
    write.csv(cons, args$out, row.names = FALSE)
    cat(sprintf("Wrote %s: %d rows (one per game)\n", args$out, nrow(cons)))
  } else {
    cons <- consensus(long_df)
    write.csv(cons, args$out, row.names = FALSE)
    cat(sprintf("Wrote %s: %d rows (one per game/player/market)\n", args$out, nrow(cons)))
  }
  cat("Upload that CSV back -- no need to upload hist_raw*/ itself.\n")
} else {
  cat("No props flattened (no events matched, or all games were already empty).\n")
}

File: historical_pull.py
#!/usr/bin/env python3
"""
historical_pull.py — pull 2025 CFB player props from The Odds API historical
endpoint for a season-long backtest.

COST MODEL (from the docs, non-negotiable):
  historical event-odds = 10 credits  x  regions  x  markets  x  event  x  snapshot
  With 6 markets, 1 region, 2 snapshots/game => 120 credits per game.
  ~750 games => ~90,000 credits.

This script is built to NOT waste that budget:
  * --dry-run counts games and prints projected credits BEFORE spending anything.
  * Every response is checkpointed to disk; a completed game is never re-pulled.
    Crash at week 9 -> rerun resumes, doesn't re-spend.
  * A hard CREDIT_CEILING stops the run if projected spend exceeds it.
  * The live x-requests-remaining header is checked after each call; if it drops
    below a floor, the run halts gracefully with progress saved.

WORKFLOW PER WEEK:
  1. historical events endpoint (cheap: 1 credit) -> event ids + commence times
     for a snapshot timestamp near that week's games.
  2. for each event, pull historical event-odds at TWO timestamps:
       - opening : earliest props snapshot (~36h before commence)
       - closing : last snapshot before commence_time
  3. write raw json per (event, snapshot) to disk; flatten later.

Snapshots for CFB props: books post props ~1-2 days out, so "opening" here means
earliest-available near T-36h and "closing" means the snapshot just before kick.
The API returns the closest snapshot <= the requested timestamp.
"""
import os, json, time, argparse, datetime as dt
import urllib.request, urllib.parse, urllib.error

# ----------------- CONFIG -----------------
# Set via .env (gitignored) -- see .env.example. No hardcoded fallback.
API_KEY   = os.environ.get("ODDS_API_KEY")
SPORT     = "americanfootball_ncaaf"
REGION    = "us"
MARKETS   = ["player_pass_yds", "player_pass_attempts",
             "player_rush_yds", "player_rush_attempts",
             "player_reception_yds", "player_receptions"]
ODDS_FMT  = "american"

HIST_COST_PER = 10                      # credits per region per market per event per snapshot
CREDIT_CEILING = 100000                 # hard stop: never let a run exceed this
CREDIT_FLOOR   = 2000                   # halt if remaining drops below this
BASE = "https://api.the-odds-api.com/v4"

CKPT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "hist_raw")
# Snapshots to pull per game. Closing-only for the first backtest (sharpest
# lines, ~half the cost). Add "opening" back to this tuple for line-movement.
SNAPSHOTS = ("closing",)

# Hours-before-commence to target for each snapshot label.
SNAPSHOT_OFFSETS = {"opening": 36, "closing": 1}


# ----------------- HTTP -----------------
def _get(url):
    try:
        with urllib.request.urlopen(url, timeout=40) as r:
            remaining = r.headers.get("x-requests-remaining")
            used = r.headers.get("x-requests-last")
            body = json.loads(r.read().decode())
            return body, (int(remaining) if remaining is not None else None), used
    except urllib.error.HTTPError as e:
        print(f"    HTTP {e.code}: {e.read().decode()[:200]}")
        return None, None, None
    except Exception as e:
        print(f"    error: {e}")
        return None, None, None


def hist_events(date_iso):
    """Historical events endpoint (cheap) -> list of events at that snapshot."""
    q = urllib.parse.urlencode(dict(apiKey=API_KEY, date=date_iso))
    url = f"{BASE}/historical/sports/{SPORT}/events?{q}"
    body, remaining, _ = _get(url)
    if not body:
        return [], remaining
    data = body.get("data", body) if isinstance(body, dict) else body
    return data, remaining


def hist_event_odds(event_id, date_iso):
    """Historical event-odds for one event at one snapshot (the expensive call)."""
    q = urllib.parse.urlencode(dict(
        apiKey=API_KEY, date=date_iso, regions=REGION,
        markets=",".join(MARKETS), oddsFormat=ODDS_FMT))
    url = f"{BASE}/historical/sports/{SPORT}/events/{event_id}/odds?{q}"
    return _get(url)


# ----------------- CHECKPOINT -----------------
def ckpt_path(event_id, label):
    return os.path.join(CKPT_DIR, f"{event_id}_{label}.json")


def already_have(event_id, label):
    p = ckpt_path(event_id, label)
    return os.path.exists(p) and os.path.getsize(p) > 2


def save_ckpt(event_id, label, payload):
    os.makedirs(CKPT_DIR, exist_ok=True)
    with open(ckpt_path(event_id, label), "w") as f:
        json.dump(payload, f)


# ----------------- SCHEDULE -> SNAPSHOT TIMESTAMPS -----------------
# FBS conferences — props exist essentially only for FBS-vs-FBS games.
FBS_CONF = {"SEC", "Big Ten", "Big 12", "ACC", "American Athletic",
            "Mountain West", "Sun Belt", "Conference USA", "Mid-American",
            "Pac-12", "FBS Independents"}


def load_schedule(sched_csv, season=2025, fbs_only=True, both_fbs=True,
                  season_type="regular"):
    """
    Read the CFBD schedule CSV, return list of (commence_dt, home, away, week).
    fbs_only + both_fbs filter out FCS/lower-div games that carry no props,
    so we don't burn 120 credits/game checking games that return nothing.

    CFBD numbers postseason (bowl/CFP) weeks starting back at 1, distinct
    from the regular season's own Week 1 -- so `--week 1` without a
    SeasonType filter pulls both, more than doubling scope/cost.
    season_type: "regular" (default), "postseason", or "all".
    """
    import csv
    games = []
    seen_ids = set()
    with open(sched_csv, encoding="utf-8-sig") as f:
        for r in csv.DictReader(f):
            r = {k.strip().strip('"'): v for k, v in r.items()}
            gid = r.get("Id")
            if gid is not None:
                if gid in seen_ids:
                    continue  # the schedule export has dup rows per game (one per outlet)
                seen_ids.add(gid)
            if str(r.get("Season")) != str(season):
                continue
            if season_type != "all" and r.get("SeasonType") != season_type:
                continue
            if fbs_only:
                h_fbs = r.get("HomeConference") in FBS_CONF
                a_fbs = r.get("AwayConference") in FBS_CONF
                if both_fbs and not (h_fbs and a_fbs):
                    continue
                if not both_fbs and not (h_fbs or a_fbs):
                    continue
            st = r.get("StartTime", "")
            if not st:
                continue
            try:
                cdt = dt.datetime.fromisoformat(st.replace("Z", "+00:00"))
            except ValueError:
                continue
            games.append((cdt, r.get("HomeTeam", ""), r.get("AwayTeam", ""), r.get("Week", "")))
    return games


def snapshot_iso(commence_dt, label):
    off = SNAPSHOT_OFFSETS[label]
    ts = commence_dt - dt.timedelta(hours=off)
    return ts.astimezone(dt.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


# ----------------- MAIN -----------------
def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--schedule", required=True, help="2025 CFBD schedule CSV")
    ap.add_argument("--season", type=int, default=2025)
    ap.add_argument("--dry-run", action="store_true",
                    help="count games + projected credits, spend nothing")
    ap.add_argument("--max-games", type=int, default=None,
                    help="cap games this run (for staged spending)")
    ap.add_argument("--week", type=int, default=None, help="restrict to one week")
    ap.add_argument("--include-one-fbs", action="store_true",
                    help="include games where only ONE team is FBS (more games, more credits)")
    ap.add_argument("--season-type", default="regular",
                    choices=["regular", "postseason", "all"],
                    help="CFBD SeasonType filter (default: regular). Postseason "
                         "weeks are numbered separately from the regular season's, "
                         "so --week without this pulls both.")
    args = ap.parse_args()

    games = load_schedule(args.schedule, args.season,
                          fbs_only=True, both_fbs=not args.include_one_fbs,
                          season_type=args.season_type)
    if args.week is not None:
        games = [g for g in games if str(g[3]) == str(args.week)]
    games.sort(key=lambda g: g[0])
    if args.max_games:
        games = games[:args.max_games]

    n_games = len(games)
    n_snaps = len(SNAPSHOTS)
    projected = n_games * n_snaps * len(MARKETS) * HIST_COST_PER

    print(f"Season {args.season}: {n_games} games in scope")
    print(f"Markets: {len(MARKETS)} | snapshots/game: {n_snaps} | region: {REGION}")
    print(f"Projected MAX credits (if none cached): "
          f"{n_games} x {n_snaps} x {len(MARKETS)} x {HIST_COST_PER} = {projected:,}")

    # Count snapshot files already on disk. This can't be matched to *this*
    # scope's games without hitting the (cheap) events endpoint to discover
    # event ids first, and dry-run spends zero credits by design -- so it's
    # a general "how much is already checkpointed" figure, not a precise
    # count of what this run would skip.
    cached = 0
    if os.path.isdir(CKPT_DIR):
        cached = sum(1 for fn in os.listdir(CKPT_DIR)
                     if any(fn.endswith(f"_{lbl}.json") for lbl in SNAPSHOTS))
    print(f"Checkpoint dir: {CKPT_DIR} ({cached} snapshot file(s) already on disk)")

    if projected > CREDIT_CEILING:
        print(f"\n*** PROJECTED {projected:,} EXCEEDS CEILING {CREDIT_CEILING:,} ***")
        print("Reduce scope (--max-games or --week) or raise CREDIT_CEILING deliberately.")
        if not args.dry_run:
            return

    if args.dry_run:
        print("\n[dry-run] No credits spent. Re-run without --dry-run to execute.")
        # Show a few sample snapshot timestamps so you can eyeball them
        for cdt, h, a, wk in games[:3]:
            print(f"  wk{wk} {a} @ {h} {cdt.isoformat()}")
            for lbl in SNAPSHOTS:
                print(f"      {lbl:8s} -> {snapshot_iso(cdt, lbl)}")
        return

    # ---- EXECUTE ----
    os.makedirs(CKPT_DIR, exist_ok=True)
    spent_calls = 0
    for i, (cdt, home, away, wk) in enumerate(games, 1):
        # discover event id near this game's closing snapshot
        disc_iso = snapshot_iso(cdt, "closing")
        evs, remaining = hist_events(disc_iso)
        # match our game by team names in the snapshot's event list
        ev = _match_event(evs, home, away)
        if not ev:
            print(f"[{i}/{n_games}] wk{wk} {away}@{home}: no event match at {disc_iso}")
            continue
        eid = ev["id"]

        for lbl in SNAPSHOTS:
            if already_have(eid, lbl):
                continue
            body, remaining, used = hist_event_odds(eid, snapshot_iso(cdt, lbl))
            if body is not None:
                save_ckpt(eid, lbl, body)
                spent_calls += 1
            if remaining is not None and remaining < CREDIT_FLOOR:
                print(f"\n*** remaining credits {remaining} < floor {CREDIT_FLOOR}. "
                      f"Halting with progress saved. ***")
                return
            time.sleep(0.25)

        if i % 10 == 0:
            print(f"[{i}/{n_games}] wk{wk} done | remaining credits: {remaining}")

    print(f"\nComplete. Pulled {spent_calls} new (event,snapshot) odds files into {CKPT_DIR}")


def _norm(s):
    """Normalize for matching: strip accents, lowercase, keep alphanumerics."""
    import unicodedata
    s = unicodedata.normalize("NFKD", s or "")
    s = "".join(c for c in s if not unicodedata.combining(c))  # drop accents (é->e)
    return "".join(c for c in s.lower() if c.isalnum())


def load_team_map(path="team_map.csv"):
    """cfbd_name -> normalized odds_name, for explicit overrides."""
    import csv, os
    m = {}
    if os.path.exists(path):
        for r in csv.DictReader(open(path, encoding="utf-8-sig")):
            m[_norm(r["cfbd_name"])] = _norm(r["odds_name"])
    return m


_TEAM_MAP = load_team_map()


def _match_event(events, home, away):
    """
    Match a schedule game to an API event by team names.
    Uses accent-stripped normalization + an explicit crosswalk (team_map.csv)
    for names the fuzzy match can't bridge (App State -> Appalachian State, etc.).
    """
    # apply explicit override if present, else normalize the schedule name
    h = _TEAM_MAP.get(_norm(home), _norm(home))
    a = _TEAM_MAP.get(_norm(away), _norm(away))
    for e in events:
        eh, ea = _norm(e.get("home_team")), _norm(e.get("away_team"))
        # match orientation-agnostically: schedule and API sometimes disagree on
        # which team is home (neutral sites, data-entry differences), so accept
        # either arrangement as long as both teams are present.
        fwd = (h in eh or eh in h) and (a in ea or ea in a)
        rev = (h in ea or ea in h) and (a in eh or eh in a)
        if fwd or rev:
            return e
    return None


if __name__ == "__main__":
    main()
